### PR TITLE
Add Ticket Tests (RIPD-1328):

### PR DESCRIPTION
**Pull request by mellery451**
Migrate ticket-test.js to cpp. Add more coverage for Change/Cancel ticket
handlers.

### CODE REVIEW

**Comment by HowardHinnant** [comment]
<chuckle> I rather liked the init lists. :-)  But I don't think it is a big deal either way.

**Comment by mellery451** [comment]
@HowardHinnant yeah, me too...but gcc disagreed (on linux). If you can tell me how to craft the right syntax for all compilers, I'm happy to change it.

**Comment by HowardHinnant** [comment]
Oh, wow.  Usually it is VS that does things like that.  But none of the compilers are perfect.  I think I'd rather just stick with `make_tuple` rather than wrestle with gcc. ;-)

File: src/test/app/Ticket_test.cpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    void testTicketNotEnabled ()
+    {
+        testcase ("Feature Not Enabled");
+
+        using namespace test::jtx;
+        Env env {*this};
+
+        env (ticket::create (env.master), ter(temDISABLED));
+
+        auto jv = ticket::create (env.master);
```
Comment (bachase) [inline src/test/app/Ticket_test.cpp]: I made https://github.com/bachase/rippled/commit/bebbd2c5edc6ec72baf63da8ac19b49d67c39fe8 if you want to cherrypick.

**Comment by bachase** [comment]
I prefer initializer lists too.  I'll take it for a spin in gcc and see if I spot anything.

File: src/test/app/Ticket_test.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    static auto constexpr idOne =
+      "00000000000000000000000000000000"
+      "00000000000000000000000000000001";
+
+    /// @brief validate metadata for a create/cancel ticket transaction and
+    /// return the 3 or 4 nodes that make-up the metadata (AffectedNodes)
+    ///
+    /// @param env current jtx env (meta will be extracted from it)
+    ///
+    /// @param is_cancel flag to indicate whether the latest tx was a ticket
```
Comment (mellery451) [inline src/test/app/Ticket_test.cpp]: good point - will fix.

File: src/test/app/Ticket_test.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    /// create or ticket cancel: this determines the some of the node names
+    /// that are extracted
+    ///
+    /// @param expiration flag to indicate a cancelation with expiration which
```
Comment (mellery451) [inline src/test/app/Ticket_test.cpp]: will fix

File: src/test/app/Ticket_test.cpp (hunk 1 above)
Comment (mellery451) [inline src/test/app/Ticket_test.cpp]: looks good  - thx.

**Comment by HowardHinnant** [comment]
👍  lgtm

File: src/test/app/Ticket_test.cpp [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+        auto const& jdir = crd[2];
+        BEAST_EXPECT(jdir[sfLedgerIndex.fieldName] ==
+            jticket[sfLedgerIndex.fieldName]);
+        BEAST_EXPECT(jdir[sfFinalFields.fieldName][jss::Account] ==
+            env.master.human());
+        BEAST_EXPECT(jdir[sfFinalFields.fieldName][sfTarget.fieldName] ==
+            alice.human());
+        BEAST_EXPECT(jdir[sfFinalFields.fieldName][jss::Flags] == 0);
+        BEAST_EXPECT(jdir[sfFinalFields.fieldName][sfOwnerNode.fieldName] ==
+            "0000000000000000");
```
Comment (mellery451) [inline src/test/app/Ticket_test.cpp]: yes - I will add that check to make it explicit

**Comment by bachase** [comment]
Looks like @HowardHinnant already answered the initializer list question http://stackoverflow.com/questions/14961809/returning-a-tuple-from-a-function-using-uniform-initialization-syntax ;)

**Comment by HowardHinnant** [comment]
Ha! My mouth is so big I can't be expected to remember everything I say! :-)

File: src/test/app/Ticket_test.cpp [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+        Env env {*this, features (featureTickets)};
+
+        env.timeKeeper ().adjustCloseTime (days {2});
+        env.close ();
+
+        // create and verify
+        uint32_t expire = 60;
+        env (ticket::create (env.master, expire));
+        // in the case of past expiration, we only get
+        // one meta node entry returned
```
Comment (mellery451) [inline src/test/app/Ticket_test.cpp]: well, every tx charges a fee...I think the real point here is that a ticket is not created...so we have no CreatedNode entry

Comment (bachase) [inline src/test/app/Ticket_test.cpp]: Got it. Thanks

**Comment by nbougalis** [comment]
Merged as 7e39d645b9f481b6142b32b6d295f06531dc51d1

**Pull request closed without merging**